Add Constant Dynamic support in GC

fengxue-IS opened the pull request:
Update GC_ConstantPoolObjectSlotIterator to recognize constantDynamic

Close: #1273

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
depends: #1426

dmitripivkine commented on runtime/gc_structs/ConstantPoolObjectSlotIterator.cpp:
```diff
@@ -73,6 +73,13 @@ GC_ConstantPoolObjectSlotIterator::nextSlot()
 			return &(((J9RAMMethodHandleRef *) slotPtr)->methodHandle);
 		}
+		if(slotType == J9CPTYPE_CONSTANT_DYNAMIC) {
+			if(((J9RAMConstantDynamicRef *) slotPtr)->value != NULL) {
```
Would you please put NULL to the left side:
if(NULL != ((J9RAMConstantDynamicRef *) slotPtr)->value) {

dmitripivkine commented on runtime/gc_structs/ConstantPoolObjectSlotIterator.cpp:
```diff
@@ -73,6 +73,13 @@ GC_ConstantPoolObjectSlotIterator::nextSlot()
 		if(slotType == J9CPTYPE_METHODHANDLE) {
 			return &(((J9RAMMethodHandleRef *) slotPtr)->methodHandle);
 		}
+		if(slotType == J9CPTYPE_CONSTANT_DYNAMIC) {
```
Would you mind to replace all if's to the switch? I believe it will work faster and look cleaner

dmitripivkine commented:
This PR can be tested and submitted when PRs it depends at are ready and in.

DanHeidinga commented:
Does this change actually depend on #1533?  Or is possible to merge it before the resolve code is there as the GC should just see both slots as null until we resolve one.

fengxue-IS commented:
@DanHeidinga this doesn't depend on the code change from #1533, I added that as a dependency since this will only be used after #1533 has been merged
fixed compile issue, testing

DanHeidinga commented:
@dmitripivkine This has changed slightly since you last reviewed it.  Do you want to give it another look?

dmitripivkine commented:
jenkins compile pLinux jdk8

dmitripivkine commented:
Jenkins test sanity aix jdk9
